Implement CSS Values 5 random()

weinig opened the pull request:
#### b3ba78888598c43590405267ac1d2c740f6e3cdb
<pre>
Implement CSS Values 5 random()
<a href="https://bugs.webkit.org/show_bug.cgi?id=284344">https://bugs.webkit.org/show_bug.cgi?id=284344</a>

Reviewed by Darin Adler.

Implements CSS Values 5 random() math function behind a
new feature flag, CSSRandomFunctionEnabled.

random() requires maintaining state on a per-document basis for
random values based on a &quot;random cache key&quot; derived from the
parameters to random(). Due to the layering of the Calculation
tree in platform, the mapping was divided into two parts: a
per-element/document wide mapping and then a mapping for the
remaining bits (optional identifier, min, max, optional step).

A conservative approach was take to disable style sharing as
necessary for per-element unique values where we use the
existence of any random key map on an element as indicator
that the element&apos;s style should not be shared. This is checked
in MatchedDeclarationsCache and SharingResolver.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-computed.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-computed.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-invalid.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-invalid.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-serialize.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-serialize.tentative.html: Added.
    - Add new tests.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
    - Add feature flag.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/Headers.cmake:
    - Add new files.
* Source/WebCore/css/CSSValueKeywords.in:
    - Add new keywords.
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree+Traversal.h:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
* Source/WebCore/platform/calc/CalculationExecutor.h:
* Source/WebCore/platform/calc/CalculationOperator.cpp:
* Source/WebCore/platform/calc/CalculationOperator.h:
* Source/WebCore/platform/calc/CalculationRandomKey.h: Added.
* Source/WebCore/platform/calc/CalculationRandomKeyMap.h: Added.
* Source/WebCore/platform/calc/CalculationTree+Copy.cpp:
* Source/WebCore/platform/calc/CalculationTree+Evaluation.cpp:
* Source/WebCore/platform/calc/CalculationTree+Traversal.h:
* Source/WebCore/platform/calc/CalculationTree.cpp:
* Source/WebCore/platform/calc/CalculationTree.h:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleSharingResolver.cpp:

Canonical link: <a href="https://commits.webkit.org/288779@main">https://commits.webkit.org/288779@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/11e8aa6c7e6776ce3901cb7be89d9698b7270c67

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84399 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4024 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38704 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89479 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35409 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86484 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4109 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12003 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65635 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23476 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87445 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3092 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76681 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45929 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3043 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30912 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34457 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77363 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73933 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31681 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90861 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83416 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11666 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8490 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74086 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11893 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72505 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/73286 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18132 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17623 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16067 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3063 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11618 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17094 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/105835 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11467 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25541 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14943 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13240 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -149,7 +150,8 @@ void add(Hasher& hasher, const CSSParserContext& context)
         | context.cssContainerProgressFunctionEnabled       << 27
-        | (uint32_t)context.mode                            << 28; // This is multiple bits, so keep it last.
+        | context.cssRandomFunctionEnabled                  << 28
+        | (uint32_t)context.mode                            << 29; // This is multiple bits, so keep it last.
```
How many bits? We are getting perilously 32 here.

weinig replied:
It needs 3 bits, so just barely still works. Next addition will kill it.

anttijk replied:
Stuff like the Enabled bits probably doesn't even need to be part of the hash. Who cares if we get some collisions when someone flips the setting?

darinadler commented on Source/WebCore/platform/calc/CalculationExecutor.h:
```diff
@@ -487,6 +487,32 @@ template<> struct OperatorExecutor<Progress> {
+                return result - stepValue;
+            return result;
+        } else
+            return min + ((max - min) * randomUnitInterval);
```
Nicer to do this first as an early return I think.

darinadler reviewed (approved):
Looks good. I know it's just a draft because of the caching.

weinig committed:
Implement CSS Values 5 random()
https://bugs.webkit.org/show_bug.cgi?id=284344

Reviewed by Darin Adler.

Implements CSS Values 5 random() math function behind a
new feature flag, CSSRandomFunctionEnabled.

random() requires maintaining state on a per-document basis for
random values based on a "random cache key" derived from the
parameters to random(). Due to the layering of the Calculation
tree in platform, the mapping was divided into two parts: a
per-element/document wide mapping and then a mapping for the
remaining bits (optional identifier, min, max, optional step).

A conservative approach was take to disable style sharing as
necessary for per-element unique values where we use the
existence of any random key map on an element as indicator
that the element's style should not be shared. This is checked
in MatchedDeclarationsCache and SharingResolver.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-computed.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-computed.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-invalid.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-invalid.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-serialize.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-serialize.tentative.html: Added.
    - Add new tests.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
    - Add feature flag.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/Headers.cmake:
    - Add new files.
* Source/WebCore/css/CSSValueKeywords.in:
    - Add new keywords.
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree+Traversal.h:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
* Source/WebCore/platform/calc/CalculationExecutor.h:
* Source/WebCore/platform/calc/CalculationOperator.cpp:
* Source/WebCore/platform/calc/CalculationOperator.h:
* Source/WebCore/platform/calc/CalculationRandomKey.h: Added.
* Source/WebCore/platform/calc/CalculationRandomKeyMap.h: Added.
* Source/WebCore/platform/calc/CalculationTree+Copy.cpp:
* Source/WebCore/platform/calc/CalculationTree+Evaluation.cpp:
* Source/WebCore/platform/calc/CalculationTree+Traversal.h:
* Source/WebCore/platform/calc/CalculationTree.cpp:
* Source/WebCore/platform/calc/CalculationTree.h:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleSharingResolver.cpp:

Canonical link: https://commits.webkit.org/288779@main
